src/webgpu/api/validation/buffer_mapping.spec.ts

```diff
@@ -0,0 +1,763 @@
+          offset: 0,
+          size: 80 - 24 + kSizeAlignement,
+        },
+        {
+          bufferSize: 80,
+          mapOffset: 24,
+          mapSize: undefined,
+          offset: kOffsetAlignment,
+          size: 80 - 24,
+        },
+
+        // Test for a non-empty buffer fully mapped.
+        { bufferSize: 80, mapOffset: 0, mapSize: 80, offset: 0, size: 80 },
+        { bufferSize: 80, mapOffset: 0, mapSize: 80, offset: kOffsetAlignment, size: 80 },
+        { bufferSize: 80, mapOffset: 0, mapSize: 80, offset: 0, size: 80 + kSizeAlignement },
+
+        { bufferSize: 80, mapOffset: 0, mapSize: 80, offset: 40, size: 40 },
+        { bufferSize: 80, mapOffset: 0, mapSize: 80, offset: 40 + kOffsetAlignment, size: 40 },
+        { bufferSize: 80, mapOffset: 0, mapSize: 80, offset: 40, size: 40 + kSizeAlignement },
+
+        // Test for a buffer partially mapped.
+        { bufferSize: 80, mapOffset: 24, mapSize: 40, offset: 24, size: 40 },
+        { bufferSize: 80, mapOffset: 24, mapSize: 40, offset: 24 - kOffsetAlignment, size: 40 },
+        { bufferSize: 80, mapOffset: 24, mapSize: 40, offset: 24 + kOffsetAlignment, size: 40 },
+        { bufferSize: 80, mapOffset: 24, mapSize: 40, offset: 24, size: 40 + kSizeAlignement },
+
+        // Test for a partially mapped buffer with implicit size and offset for getMappedRange.
+        // - Buffer partially mapped in the middle
+        { bufferSize: 80, mapOffset: 24, mapSize: 40, offset: undefined, size: undefined },
+        { bufferSize: 80, mapOffset: 24, mapSize: 40, offset: 24, size: undefined },
+        // - Buffer partially mapped to the end
+        { bufferSize: 80, mapOffset: 24, mapSize: undefined, offset: 24, size: undefined },
+        // - Buffer partially mapped from the start
+        { bufferSize: 80, mapOffset: 0, mapSize: 64, offset: undefined, size: undefined },
+        { bufferSize: 80, mapOffset: 0, mapSize: 64, offset: undefined, size: 64 },
+      ])
+  )
+  .fn(async t => {
+    const { mapMode, bufferSize, mapOffset, mapSize, offset, size } = t.params;
+    const buffer = t.createMappableBuffer(mapMode, bufferSize);
+    await buffer.mapAsync(mapMode, mapOffset, mapSize);
+
+    const actualMapOffset = mapOffset ? mapOffset : 0;
+    const actualMapSize = mapSize ? mapSize : bufferSize - actualMapOffset;
+
+    const actualOffset = offset ? offset : 0;
+
+    let success = actualOffset >= actualMapOffset && actualOffset <= bufferSize;
+    if (size !== undefined) {
+      success = success && actualOffset + size <= actualMapOffset + actualMapSize;
+    }
+
+    t.testGetMappedRangeCall(success, buffer, offset, size);
+  });
+
+g.test('getMappedRange_disjointRanges')
+  .desc('Test that the ranges asked through getMappedRange must be disjoint.')
+  .params(
+    params()
+      .combine(pbool('remapBetweenCalls'))
+      .combine([
+        // Disjoint ranges with one that's empty.
+        { offset1: 8, size1: 0, offset2: 8, size2: 8 },
```

Comment:
Wouldn't that make the test more verbose because `offset1` would need to become `ranges[0].offset`? Can typescript don't seem to be able to destruct-assign sequences of dict?

I was thinking of focusing this test on only two ranges so that the logic can stay somewhat simple (and not do loops of loops)

Reply:
You can nest destructurings `const { ranges: [{ offset: offset1, size: size1 }, { offset: offset2, size: size2 }] } = t.params;`
but I was thinking about doing it with a loop. That would be overly complex if you don't want to do more than 2 ranges.